cypress/e2e/middleware/preview.cy.ts

```diff
@@ -0,0 +1,34 @@
+describe('Preview Mode', () => {
+  it('enters and exits preview mode', () => {
+    Cypress.Cookies.debug(true)
+    cy.getCookies().then((cookie) => cy.log('cookies', cookie))
+
+    cy.intercept('/previewTest', (req) => {
+      req.continue((res) => {
+        expect(res.headers?.['x-middleware-executed']).to.equal('true')
+      })
+    }).as('previewTestVisit')
+
+    // preview mode is off by default
+    cy.visit('/previewTest')
+    cy.findByText('Is preview? No', { selector: 'h1' })
+
+    // enter preview mode
+    cy.request('/api/enterPreview').then((response) => {
+      expect(response.body).to.have.property('name', 'preview mode')
+    })
+
+    // exptected content is rendered
+    cy.visit('/previewTest')
+    cy.findByText('Is preview? Yes!', { selector: 'h1' })
+
+    // exit preview mode
+    cy.request('/api/exitPreview')
+    cy.visit('/previewTest')
+    cy.findByText('Is preview? No', { selector: 'h1' })
+
+    // we should hit /previewTest 3 times (before entering preview, after entering preview, after exiting preview)
+    // this assertion is mainly to ensure interception works and assertion on response header is made
+    cy.get('@previewTestVisit.all').should('have.length', 3)
+  })
+})
```

Comment:
This test does assert things correctly, but the version of Next used in `middleware` demo actually doesn't have the bug included, so the fix is only relevant to older next versions which we don't run test against currently :/

I think it's still worthwhile to include the test, but we should figure out test setup to run against more versions of next.js

Reply:
It's good to have this like you mentioned and if we get around to running these tests on older versions, this will definitely be helpful.